api/cron/prune_models.py

```diff
@@ -33,6 +33,211 @@
 from build_config import build_config  # noqa isort:skip
 
 
+class ModelPruner:
+    def __init__(
+        self, region, endpoint_last_used_cutoff, test_run, endpoints_name_not_touch
+    ):
+        self.region = region
+        self.endpoint_last_used_cutoff = endpoint_last_used_cutoff
+        self.test_run = test_run
+        self.endpoints_name_not_touch = endpoints_name_not_touch
+
+        # Setup AWS session
+        session = boto3.Session(
+            aws_access_key_id=build_config["aws_access_key_id"],
+            aws_secret_access_key=build_config["aws_secret_access_key"],
+            region_name=region,
+        )
+
+        self.sm = session.client("sagemaker")
+        self.cw = session.client("cloudwatch")
+
+        self.obj_type_to_sm_method = {
+            "endpoint": self.sm.list_endpoints,
+            "sagemaker_model": self.sm.list_models,
+            "endpoint_config": self.sm.list_endpoint_configs,
+        }
+
+        self.obj_type_to_sm_resp_field = {
+            "endpoint": "Endpoints",
+            "sagemaker_model": "Models",
+            "endpoint_config": "EndpointConfigs",
+        }
+
+        self.endpoints_deleted = []
+
+    def _dont_touch_endpoint(self, ep_name):
+        return len(self.endpoints_name_not_touch) > 0 and (
+            any(
+                ep_name in bad_endpoint
+                for bad_endpoint in self.endpoints_name_not_touch
+            )
+        )
+
+    def _delete_from_aws(self, endpoint_name):
+        print(f"Removing model endpoint at: {endpoint_name}")
+
+        # Delete Endpoint
+        endpoint_to_delete = self._find_obj(endpoint_name, "endpoint")
+        if endpoint_to_delete is not None:
+            self.sm.delete_endpoint(EndpointName=endpoint_name)
+
+        # Delete Sagemaker model
+        model_to_delete = self._find_obj(endpoint_name, "sagemaker_model")
+        if model_to_delete is not None:
+            self.sm.delete_model(ModelName=endpoint_name)
+
+        # Delete Endpoint Config
+        endpoint_config_to_delete = self._find_obj(endpoint_name, "endpoint_config")
+        if endpoint_config_to_delete is not None:
+            self.sm.delete_endpoint_config(EndpointConfigName=endpoint_name)
+
+    def _find_obj(self, ep_name, obj_name):
+        sm_method = self.obj_type_to_sm_method[obj_name]
+        sm_resp_field = self.obj_type_to_sm_resp_field[obj_name]
+
+        endpoint_response = sm_method(
+            NameContains=ep_name,
+            MaxResults=100,
+        )
+
+        endpoints = endpoint_response[sm_resp_field]
```

Comment:
This is what we were not doing before i.e using `nextToken` to search through all AWS objects when finding an object to delete (whether it was an endpoint, sagemaker model, or endpoint config)